include/oneapi/dpl/experimental/kt/esimd_radix_sort_utils.h

```diff
@@ -3,79 +3,84 @@
-copy_to(const SIMD& simd, Output& output, uint32_t offset)
+template <typename T, int N>
+void
+copy_to(T* output, ::std::uint32_t base_offset, const sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_to(output + offset);
+    values.copy_to(output + base_offset);
 }
 
-template<typename SIMD, typename Output>
-inline
-typename ::std::enable_if_t<!::std::is_pointer_v<Output>, void>
-copy_to(const SIMD& simd, Output& output, uint32_t offset)
+template <typename AccessorT, typename T, int N, typename... Args>
+typename ::std::enable_if<!::std::is_pointer<AccessorT>::value, void>::type
+copy_to(AccessorT& output, ::std::uint32_t base_offset, const sycl::ext::intel::esimd::simd<T, N>& values)
 {
-    simd.copy_to(output, offset);
+    values.copy_to(output, base_offset * sizeof(T));
 }
 
-template<typename T, int N, typename InputT>
-inline
-typename ::std::enable_if_t<::std::is_pointer_v<InputT>, sycl::ext::intel::esimd::simd<T, N>>
-gather(const InputT& input, sycl::ext::intel::esimd::simd<T, N> offsets, uint32_t base_offset)
+template <typename T, int N>
+sycl::ext::intel::esimd::simd<T, N>
+gather(const T* input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets, ::std::uint32_t base_offset)
 {
-    return sycl::ext::intel::esimd::gather(input + base_offset, offsets*static_cast<uint32_t>(sizeof(T)));
+    //static_assert(::std::is_same<decltype(input), decltype(input + base_offset)>::value, "'input' and 'input + base_offset' are not the same");
+    //static_assert(::std::is_same<decltype(offsets), decltype(offsets * sizeof(T))>::value, "'offsets' and 'offsets * sizeof(T)' are not the same");
+    //decltype(offsets)::dummy;
+    //decltype(sizeof(T))::dummy;
+    //decltype(offsets * sizeof(T))::dummy;
+    return sycl::ext::intel::esimd::gather(input + base_offset, offsets/* * sizeof(T)*/);
 }
 
-template<typename T, int N, typename InputT>
-inline
-typename ::std::enable_if_t<!std::is_pointer_v<InputT>, sycl::ext::intel::esimd::simd<T, N>>
-gather(const InputT& input, sycl::ext::intel::esimd::simd<T, N> offsets, uint32_t base_offset)
+template <typename T, int N, typename AccessorT>
+typename ::std::enable_if<!::std::is_pointer<AccessorT>::value, sycl::ext::intel::esimd::simd<T, N>>::type
+gather(AccessorT input, sycl::ext::intel::esimd::simd<::std::uint32_t, N> offsets,::std::uint32_t base_offset)
 {
-    return sycl::ext::intel::esimd::gather<T>(input, offsets*static_cast<uint32_t>(sizeof(T)),
-                                              base_offset*static_cast<uint32_t>(sizeof(T)));
+    //static_assert(::std::is_same<decltype(offsets), decltype(offsets * sizeof(T))>::value, "'offsets' and 'offsets * sizeof(T)' are not the same");
+    //static_assert(::std::is_same<decltype(base_offset), decltype(base_offset * sizeof(T))>::value, "'base_offset' and 'base_offset * sizeof(T)' are not the same");
+
+    return sycl::ext::intel::esimd::gather<T>(input, offsets /** sizeof(T)*/, base_offset /** sizeof(T)*/);
```

Comment:
The comment is the same as in https://github.com/oneapi-src/oneDPL/pull/871/files#r1156252968, but the imitated API is [that](https://intel.github.io/llvm-docs/doxygen/group__sycl__esimd__memory.html#gab8e33bce282903e5c4e74b009ed0d0c9). Both offsets should be multiplied by `sizeof(T)` because `gather` expect offsets expressed in bytes. Static cast is needed to explicitly pass arguments with types which ESIMD expects.
```suggestion
    return sycl::ext::intel::esimd::gather<T>(input, offsets * static_cast<::std::uint32_t>(sizeof(T)), base_offset * static_cast<::std::uint32_t>(sizeof(T)));
```

Reply:
Fixed